modules/jod-ohjaaja-cms-study-program-importer/src/main/java/fi/okm/jod/ohjaaja/cms/studyprogram/service/StudyProgramStructureService.java

```diff
@@ -76,111 +81,352 @@ public void deleteDDMStructure() throws StudyProgramDDMStructureDeleteException
               + EXTERNAL_REFERENCE_CODE);
       return;
     }
+
     try {
-      var structureLayout =
-          ddmStructureLayoutLocalService.fetchStructureLayout(
-              JOD_GROUP_ID, structure.getClassNameId(), structure.getStructureKey());
-      if (structureLayout != null) {
-        ddmStructureLayoutLocalService.deleteDDMStructureLayout(structureLayout);
-        log.info("Deleted DDM structure layout for: " + EXTERNAL_REFERENCE_CODE);
-      }
+      long jaClassNameId = PortalUtil.getClassNameId(JournalArticle.class.getName());
+
+      deDataDefinitionFieldLinkLocalService.deleteDEDataDefinitionFieldLinks(
+          jaClassNameId, structure.getStructureId());
+
+      unlinkFieldsetFromStructure(structure);
+
+      deleteLayoutIfExists(
+          structure.getGroupId(), structure.getClassNameId(), structure.getStructureKey());
+
       ddmStructureLocalService.deleteStructure(structure);
       log.info("Deleted DDM structure: " + EXTERNAL_REFERENCE_CODE);
+
+      DDMStructure fieldSet =
+          ddmStructureLocalService.fetchStructure(JOD_GROUP_ID, jaClassNameId, FIELDSET_KEY);
+
+      if (fieldSet != null) {
+        deDataDefinitionFieldLinkLocalService.deleteDEDataDefinitionFieldLinks(
+            jaClassNameId,
+            structure.getStructureId()); // varmistus; ei haittaa vaikka ei olisi rivejä
+
+        deleteLayoutIfExists(
+            fieldSet.getGroupId(), fieldSet.getClassNameId(), fieldSet.getStructureKey());
+
+        ddmStructureLocalService.deleteStructure(fieldSet);
+        log.info("Deleted field set: " + FIELDSET_KEY);
+      }
     } catch (PortalException e) {
       log.error("Failed to delete DDM structure: " + EXTERNAL_REFERENCE_CODE, e);
       throw new StudyProgramDDMStructureDeleteException(
           "Failed to delete DDM structure: " + EXTERNAL_REFERENCE_CODE);
     }
   }
 
-  private DDMStructure createDDMStructure(long classNameId) throws PortalException {
+  private void unlinkFieldsetFromStructure(DDMStructure main) throws PortalException {
+
+    var form = main.getDDMForm();
+    var kept = new ArrayList<DDMFormField>();
+    for (var f : form.getDDMFormFields()) {
+      if (!FIELDSET_FIELD_NAME.equals(f.getName())) {
+        kept.add(f);
+      }
+    }
+    form.setDDMFormFields(kept);
+
+    var oldLayout = main.getDDMFormLayout();
+    var newLayout = new DDMFormLayout();
+    newLayout.setDefaultLocale(oldLayout.getDefaultLocale());
+    newLayout.setPaginationMode(oldLayout.getPaginationMode());
+
+    var newPages =
+        oldLayout.getDDMFormLayoutPages().stream()
+            .map(
+                page -> {
+                  var newRows =
+                      page.getDDMFormLayoutRows().stream()
+                          .map(
+                              row -> {
+                                var newCols =
+                                    row.getDDMFormLayoutColumns().stream()
+                                        .map(
```

Comment:
using flatmap and returing a (possibly empty) stream would avoid the "filter nulls".